File: problem/SingleCellGrow/PMV/R/simulator1.R
simulator1 <-
function(T, r1, r2, d1, d2, p, w){
  ## x is the system state
  x <- c(0, 0)

  ## define reaction system
  nr <- 5
  ns <- 2
  c <- c(r1, r2, d1, d2, p)
  nu <- matrix(0, ns, nr)
  nu[,1] = c(1, 0)
  nu[,2] = c(0, 1)
  nu[,3] = c(-1, 0)
  nu[,4] = c(0, -1)
  nu[,5] = c(-1,1)
  
  ## determing initial cell
  u <- runif(1)
  if ( u < w ) {
    x[1] = 1
  } else {
    x[2] = 1
  }

  ## Begin main loop
  .t <- 0
  while ( TRUE ){
    a <- getrate(x, c)
    if ( sum(a) == 0 ) break
    ## cat(a, '\n')
    ca <- cumsum(a)
    tau <- rexp(1, rate = 1)
    tau <- tau/ca[nr]
    .t <- .t + tau
    ## cat(.t, '\n')
    if ( .t > T ) {
      ## cat(.t, x, '\n')
      ## cat('breaked')
      break
    }
    
    u <- runif(1)
    u <- u*ca[nr]
    rindex <- -1
    for ( j in seq(nr) ){
      if ( u < ca[j] ) {
        rindex <- j
        break
      }
    }
    x <- x + nu[,rindex]
    ## cat(x, '\n')
  }
  return(x)
}

File: trunk/problem/SingleCellGrow/PMV/src/simulator1.c
#include <R.h>
#include <Rmath.h>


void getrate(int *x, double *c, double *a){
  a[0] = c[0]*x[0];
  a[1] = c[1]*x[1];
  a[2] = c[2]*x[0];
  a[3] = c[3]*x[1];
  a[4] = c[4]*x[0];
}

void simulator1(double *Tin, double *l1in, double *l2in, double *d1in,
		double *d2in, double *vin, double *win, int *x){
  double T = Tin[0];
  double l1 = l1in[0];
  double l2 = l2in[0];
  double d1 = d1in[0];
  double d2 = d2in[0];
  double v = vin[0];
  double w = win[0];  

  int nr = 5;
  int ns = 2;
  double c[5] = {l1, l2, d1, d2, v};
  double a[5];
  double ca[5];
  int nu[5][2] = {{1,0}, {0,1}, {-1,0}, {0,-1}, {-1,1}};
  
  int rindex, j;
  double u;
  double t_, tau;

  GetRNGstate();

  /* determing initial cell */
  u = unif_rand();
  if ( u < w ) {
    x[0] = 1;
  }
  else {
    x[1] = 1;
  }

  /* Begin main loop */
  t_ = 0.0;
  while ( TRUE ){
    getrate(x, c, a);
    ca[0] = a[0];

    for (j=1; j<nr; j++)
      ca[j] = ca[j-1] + a[j];

    if ( ca[4] == 0 ) break;

    tau = exp_rand();
    tau = tau/ca[4];
    t_ = t_ + tau;
    
    if ( t_ > T ) {
      break;
    }
    
    u = unif_rand();
    u = u*ca[4];
    rindex = -1;
    for ( j=0; j<5; j++ ){
      if ( u < ca[j] ) {
        rindex = j;
	break;
      }
    }
    x[0] = x[0] + nu[rindex][0];
    x[1] = x[1] + nu[rindex][1];
  }

  PutRNGstate();
}

